src/common/exp-syntax.k

```diff
@@ -8,60 +8,84 @@ module EXP-SYNTAX
-//syntax MethodSpec ::= MethodName [klabel('Method)]
-//                    |Exp "." K Id [klabel('Method)]
-//                    |"super" "." K Id [klabel('SuperMethod)]
-//                    |TypeName "." "super" "." K Id [klabel('QSuperMethod)]
-//                    |AmbName "." K Id [klabel('GenericMethod)]
+syntax StmtExp ::= PrefixPostfixExp|AssignExp|MethodInvokeExp|ClassInstanceCreationExp
 
-syntax MethodName ::= Id | MethodRef
-syntax MethodInvokeExp ::= K "." MethodName "(" Exps ")"       [seqstrict(1,3), klabel('MethodImpl)]
+//@ \subsection{MethodInvocation.sdf}
+syntax MethodInvokeExp ::= MethodSpec "(" Exps ")"          [klabel('Invoke)]
+syntax MethodSpec ::= MethodName                            [klabel('Method)]
+                    | Exp "." TypeArgs Id                   [prefer, klabel('Method)]
+                    | Exp "." Id                            [klabel('Method)]
+                    | "super" "." TypeArgs Id               [prefer, klabel('SuperMethod)]
+                    | "super" "." Id                        [klabel('SuperMethod)]
+                    | TypeName "." "super" "." TypeArgs Id  [prefer, klabel('QSuperMethod)]
+                    | TypeName "." "super" "." Id           [klabel('QSuperMethod)]
+                    | AmbName "." TypeArgs Id               [klabel('GenericMethod)]
+
+//Denis' syntax
+syntax MethodInvokeExp ::= K "." MethodName "(" Exps ")"    [seqstrict(1,3), klabel('MethodImpl)]
+
+//@ \subsection{ClassInstanceCreation.sdf}
+/*todo optionals, I usually used two syntax and one with [prefer] attribute.
+But ClassInstanceCreationExp has two or three optionals inside one syntax declaration,
+I feel I am doing wired things.*/
+syntax ClassInstanceCreationExp ::= "new" TypeArgs ClassOrInterfaceType "(" Exps ")" ClassBody  [klabel('NewInstance)]
+                                  | "new" ClassOrInterfaceType "(" Exps ")" ClassBody  [klabel('NewInstance)]
+                                  | "new" TypeArgs ClassOrInterfaceType "(" Exps ")"  [klabel('NewInstance)]
+                                  | "new" ClassOrInterfaceType "(" Exps ")"  [klabel('NewInstance)]
+                                  | Exp "." "new" TypeArgs Id TypeArgs "(" Exps ")" ClassBody [klabel('QNewInstance)]
+                                  | Exp "." "new" TypeArgs Id TypeArgs "(" Exps ")"  [klabel('QNewInstance)]
+                                  | Exp "." "new" TypeArgs Id "(" Exps ")" ClassBody [klabel('QNewInstance)]
+                                  | Exp "." "new" TypeArgs Id "(" Exps ")"  [klabel('QNewInstance)]
+                                  | Exp "." "new" Id TypeArgs "(" Exps ")" ClassBody [klabel('QNewInstance)]
+                                  | Exp "." "new" Id TypeArgs "(" Exps ")" [klabel('QNewInstance)]
+                                  | Exp "." "new" Id "(" Exps ")" ClassBody [klabel('QNewInstance)]
+                                  | Exp "." "new" Id "(" Exps ")" [klabel('QNewInstance)]
```

Comment:
Again, there are lots of options above, and they all need to be desugared into the most general form (the one with `prefer`).  I guess this currently happens in JavaFront, which in my view is not right.  We should do that with macros in this syntax module, and modify JavaFront to not do the desugaring itself.  Otherwise, we will not be able to smoothly replace the JavaFront parser with our own, when ready.